rpc/src/rpc_actor.rs

```diff
@@ -134,6 +150,50 @@ impl Receive<ShellChannelMsg> for RpcServer {
 
     fn receive(&mut self, _ctx: &Context<Self::Msg>, msg: ShellChannelMsg, _sender: Sender) {
         if let ShellChannelMsg::NewCurrentHead(_, block) = msg {
+            // prepare main chain_id
+            let chain_id = parse_chain_id(MAIN_CHAIN_ID, &self.env).unwrap();
+            // warm-up - calls where chain_id + block_hash
```

Comment:
@tizoc 
and another thing is maybe we could trigger this warms in parallel with `futures::try_join!,`
I used `try_join!` in base_service e.g.

Reply:
hmmmm I tried to do that now, but it requires `receive` to be `async` too:

```
error[E0728]: `await` is only allowed inside `async` functions and blocks
```

But that trait is defined by `riker`, and it doesn't use `async`.

@bkontur but many of those functions are `async` and we are not awaiting for their result. I am not familiar with how Rust async works, but if it is like in other similar things I have used,  the `async` ones will all be launched without waiting for the previous results since nothing is being `await`ed

Reply:
Thinking about this more, if those are not awaited, maybe they are not even being executed, so it is possible that this code here is not correct.

Reply:
Ok, just pushed a workaround. One thing is that not all those calls are `async`, maybe something needs to be done about that. Used `join!` instead of `try_join!` because the later seems to cancel everything else after any of the tasks fails.

Reply:
That `block_on` will make this `receive` method block btw, but I don't know the implications of that since I am not familiar with this side of the code, so please take a look tomorrow, maybe this is not the right place to do this.

Reply:
@tizoc 
maybe instead: `futures::executor::block_on` we should move to rpc_actor and use: `tokio_executor: &Handle`

Reply:
@tizoc 
and I dont know, if it is possibility to set sometime here: `futures::executor::block_on`,
tokio has something like `tokio::time::timeout` and we use it somewhere for async/await

Reply:
Added a timeout here: https://github.com/tezedge/tezedge/pull/638/commits/12fdbe096848b93217a33e6ecff2ac532f6c31f0

I don't know if 3 seconds is too little or too much. Will wait for drone and see if the warning shows up on the logs or not.

Reply:
Ok, turns out none of this works. Python tests started failing, so I checked the logs.

I get this from the `block_on` call:

```
cannot execute `LocalPool` executor from within another executor: EnterError
```

and this when adding the timeout:

```
there is no reactor running, must be called from the context of a Tokio 1.x runtime
```

I don't know yet how to interpret this, but overall it looks like a different approach is needed to await for those futures.

Reply:
Oh, just found `tokio_runtime.block_on`, trying that.

Reply:
@bkontur something that would be need for this to work is https://docs.rs/tokio/1.7.1/tokio/runtime/struct.Handle.html#method.block_on

so that I can do:

```rust
            // warm-up - calls where chain_id + block_hash
            if is_bootstrapped {
                let tokio_runtime_handle = tokio::runtime::Handle::current();
                if let Err(err) = tokio_runtime_handle.block_on(tokio::time::timeout(
                    RPC_WARMUP_TIMEOUT,
                    warm_up_rpc_cache(&chain_id, &block, &self.env),
                )) {
                    warn!(
                        self.env.log(),
                        "RPC warmup timeout after {:?}: {:?}", RPC_WARMUP_TIMEOUT, err
                    );
                }
            }
```

But we are using 1.2.0, and that method was added to 1.4.0.

Is there anything that keeps us at 1.2.0?

Reply:
Well, I think that would not work either, I tried similar variations that are possible in 1.2.0, but the error is always something related to:

```
there is no reactor running, must be called from the context of a Tokio 1.x runtime
```

I don't fully understand how riker and tokio relate, but I guess riker actors don't run inside the tokio runtime context.

Reply:
Passing the handle from light_node's main and using that seems to work. I don't really like how unstructured it is.

```rust
                self.tokio_executor.spawn(async move {
                    tokio::time::timeout(
                        RPC_WARMUP_TIMEOUT,
                        warm_up_rpc_cache(chain_id, block, env),
                    )
                });
 ```
 
 Btw, @mambisi suggested I use `tokio::join!` instead of the version in `futures`, which I guess makes sense. Why do other parts of the code use the `futures` one?

Reply:
@tizoc 
wow, you just extended `ShellChannelMsg::NewCurrentHead` with `is_bootstrapped`, very good, very good :)

Reply:
@tizoc @mambisi 
yes, we should better use tokio::try_join/join, but we should move `tokio_executor: Handle` to RpcServiceEnvironment, and I am not sure if should also wrap it with Arc:
```
tokio_executor: Arc<Handle>
```

Reply:
@tizoc 
yes, I saw that tokio is 1.7 now, we should upgrade deps https://viablesystems.atlassian.net/browse/TE-492
but I want to do it better after release, hope fully today

Reply:
@tizoc @mambisi 
ok, I am moving now tokio_executor to env and replaceing the futures::try_join